src/proposal.mligo

```diff
@@ -277,84 +282,102 @@ let do_total_vote_meet_quorum_threshold (proposal, store: proposal * storage): b
-    in
-    let cons = fun (l, e : operation list * operation) -> e :: l in
-    let ops = List.fold cons ops new_ops in
-    let store = remove_from_proposal_sort_by_level (start_level, proposal_key, store) in
-    (ops, store, counter)
-  else (ops, store, counter)
+    : (operation list * storage) =
+  let cond = do_total_vote_meet_quorum_threshold (proposal, store)
+          && proposal.upvotes > proposal.downvotes
+  in
+  let store = unstake_proposer_token
+        (config.rejected_proposal_slash_value, cond, proposal, config.period, config.fixed_proposal_fee_in_token, store) in
+  let (new_ops, store) =
+    if cond
+    then
+      let dl_out = config.decision_lambda { proposal = proposal; extras = store.extra } in
+      let guardian = match dl_out.guardian with
+        | Some g -> g
+        | None -> store.guardian
+      in (dl_out.operations,
+            { store with extra = dl_out.extras
+            ; guardian = guardian
+            })
+    else (nil_op, store)
+  in
+  let cons = fun (l, e : operation list * operation) -> e :: l in
+  let ops = List.fold cons ops new_ops in
+  (ops, store)
+
+
+(*
+ * Flush a proposal, starting from the first. Remove it when the flush is done, and move to the next proposal.
+ *)
+let rec flush_one (counter, config, store, ops : counter * config * storage * operation list): (operation list * storage * counter) =
+  // Ensure plist is not empty.
+  match store.ongoing_proposals_dlist with
+    | Some plist ->
+        // Ensure counter is not reached yet.
+        if (counter.current < counter.total) then
+          let proposal = check_if_proposal_exist (plist.first, store) in
+
+          // Ensure the proposal is not expired.
+          let _ = if is_proposal_age (proposal, config.proposal_expired_level)
+              then (failwith expired_proposal : unit)
+              else unit in
+          // Ensure the the proposal is flushable.
+          if (is_proposal_age (proposal, config.proposal_flush_level))
+            then
+              // Unstake tokens, and execute proposal's decision_lambda.
+              let (ops, store) = handle_proposal_is_over (config, proposal, store, ops) in
+
+              // Remove the first proposal from plist.
+              let store = { store with ongoing_proposals_dlist = plist_pop store.ongoing_proposals_dlist } in
+
+              // Continue to the next proposal
+              let counter = { counter with current = counter.current + 1n } in
+              flush_one (counter, config, store, ops)
+
+            else
+              (ops, store, counter)
+        else
+          (ops, store, counter)
+    | None ->
+        (ops, store, counter)
 
-// Flush all proposals that passed their voting stage.
+(*
+ * Flush all proposals that passed their voting stage.
+ *)
 let flush(n, config, store : nat * config * storage): return =
   if n = 0n
   then (failwith empty_flush : return)
   else
     let counter : counter = { current = 0n; total = n } in
```

Comment:
:thinking: I think that perhaps this counter type worked better before these changes, but I may be wrong.

Have you tried instead to pass the `n` total to `flush_one` and reducing it by 1 at each iteration?
Stopping when it reaches `0` should have the same result.

Reply:
Make sense, I've removed counter, and just use `n` instead.